src/measure_theory/covering/besicovitch_vector_space.lean

```diff
@@ -0,0 +1,607 @@
+
+## Main definitions and results
+
+* `satellite_config α N τ` is the type of all satellite configurations of `N+1` points
+  in the metric space `α`, with parameter `τ`.
+* `has_besicovitch_covering` is a class recording that there exist `N` and `τ > 1` such that
+  there is no satellite configuration of `N+1` points with parameter `τ`. We show that
+  finite-dimensional real vector spaces satisfy this property.
+* `multiplicity E` is the maximal number of points one can put inside the unit ball
+  of radius `2` in the vector space `E`, under the condition that their distances
+  are bounded below by `1`.
+* `multiplicity_le E` shows that `multiplicity E ≤ 5 ^ (dim E)`.
+* `good_τ E` is a constant `> 1`, but close enough to `1` that satellite configurations
+  with this parameter `τ` are not worst than for `τ = 1`.
+* `is_empty_satellite_config_multiplicity` is the main theorem, saying that there are
+  no satellite configurations of `(multiplicity E) + 1` points, for the parameter `good_τ E`.
+-/
+
+universe u
+open metric set finite_dimensional measure_theory filter fin
+
+open_locale ennreal topological_space
+
+noncomputable theory
+
+/-- A satellite configuration is a configuration of `N+1` points that shows up in the inductive
+construction for the Besicovitch covering theorem. It depends on some parameter `τ ≥ 1`.
+
+This is a family of balls (indexed by `i : fin N.succ`, with center `c i` and radius `r i`) such
+that the last ball intersects all the other balls (condition `inter`),
+and given any two balls there is an order between them, ensuring that the first ball does not
+contain the center of the other one, and the radius of the second ball can not be larger than
+the radius of the first ball (up to a factor `τ`). This order corresponds to the order of choice
+in the inductive construction: otherwise, the second ball would have been chosen before.
+This is the  condition `h`.
+
+Finally, the last ball is chosen after all the other ones, meaning that `h` can be strengthened
+by keeping only one side of the alternative in `hlast`.
+-/
+@[nolint has_inhabited_instance]
+structure besicovitch.satellite_config (α : Type*) [metric_space α] (N : ℕ) (τ : ℝ) :=
+(c : fin N.succ → α)
+(r : fin N.succ → ℝ )
+(rpos : ∀ i, 0 < r i)
+(h : ∀ i j, i ≠ j → (r i ≤ dist (c i) (c j) ∧ r j ≤ τ * r i) ∨
+                    (r j ≤ dist (c j) (c i) ∧ r i ≤ τ * r j))
+(hlast : ∀ i < last N, r i ≤ dist (c i) (c (last N)) ∧ r (last N) ≤ τ * r i)
+(inter : ∀ i < last N, dist (c i) (c (last N)) ≤ r i + r (last N))
+
+/-- A metric space has the Besicovitch covering property if there exist `N` and `τ > 1` such that
+there are no satellite configuration of parameter `τ` with `N+1` points. This is the condition that
+guarantees that the Besicovitch covering theorem holds. It is satified by finite-dimensional
+real vector spaces. -/
+class has_besicovitch_covering (α : Type*) [metric_space α] : Prop :=
```

Comment:
I understand correctly, the Besicovich covering theorem (proved in the other file) doesn't make use of this class. Would it make sense to refactor this, so that the theorem does make use of this class?

General question: is this "Besicovich constant" `N` important for applications? In the theorem statement, it is "unbundled", but in this class it is hidden behind an `∃`.

I guess there could be a definition `besicovich_constant` that uses `nat.find` to extract the smallest `N` that works.

Reply:
This class is important for the next theorem (not PR-ed yet), giving a cover by disjoint balls of a subset of full measure. This is the most useful version, and for it the value of `N` is irrelevant, which is why I formulated the class like this.

Reply:
Does it make sense to move this class to the other file, and prove the Besicovitch covering theorem using it?

My impression (but please correct me) is that this file is "only" providing an instance of this class. (Are there other interesting examples?) So this file shouldn't be the place where the class is introduced. The other file seems like a more natural place.

Reply:
I can definitely move it to the other file. Yes, there are other examples, reasonable Riemannian manifolds for instance. Maybe a more natural order would be to reverse the import order, i.e., have a first file doing the Besicovitch covering theorem (and introducing the typeclass), and a second file on vector spaces providing an instance of the class. Let me try that.

Reply:
@sgouezel yes, exactly. Thanks for trying!